invokeai/app/services/invocation_cache/invocation_cache_memory.py

```diff
@@ -1,19 +1,28 @@
-from queue import Queue
+from dataclasses import dataclass, field
+from time import time
 from typing import Optional, Union
 
 from invokeai.app.invocations.baseinvocation import BaseInvocation, BaseInvocationOutput
 from invokeai.app.services.invocation_cache.invocation_cache_base import InvocationCacheBase
-from invokeai.app.services.invocation_cache.invocation_cache_common import InvocationCacheStatus
+from invokeai.app.services.invocation_cache.invocation_cache_common import InvocationCacheStatus, ThreadLock
 from invokeai.app.services.invoker import Invoker
 
+thread_lock = ThreadLock()
```

Comment:
Since `ThreadLock` uses decorators, we are forced to use a global `thread_lock` here. Global variables like this are generally considered an anti-pattern, and lead to all kinds of headaches. I commented elsewhere that I thought context managers would work better than decorators for the ThreadLock. If we make that change, then we can remove this global variable.

Reply:
The context manager would also need a global variable per lock. Not sure how you would want to do it without.

Reply:
The lock should be a member of the cache. i.e. one lock for each cache instance.